pkg/util/errors/errors.go

```diff
@@ -29,10 +29,16 @@ type Aggregate interface {
 // is itself an implementation of the error interface.  If the slice is empty,
 // this returns nil.
 func NewAggregate(errlist []error) Aggregate {
-	if len(errlist) == 0 {
+	errs := []error{}
+	for _, err := range errlist {
+		if err != nil {
```

Comment:
```
if len(certFile) > 0 {
certBytes, err := fileutil.LoadCert(certFile)
if err != nil {
errs = append(errs, err)
}
cert = string(certBytes)
}
if len(keyFile) > 0 {
keyBytes, err := fileutil.LoadKey(keyFile)
if err != nil {
errs = append(errs, err)
}
key = string(keyBytes)
}
if len(caCertFile) > 0 {
caCertBytes, err := fileutil.LoadCert(caCertFile)
if err != nil {
errs = append(errs, err)
}
caCert = string(caCertBytes)
}
if len(destCACertFile) > 0 {
destCACertBytes, err := fileutil.LoadCert(destCACertFile)
if err != nil {
errs = append(errs, err)
}
destCACert = string(destCACertBytes)
}
```

vs

```
if len(certFile) > 0 {
certBytes, err := fileutil.LoadCert(certFile)
errs = append(errs, err)
cert = string(certBytes)
}
if len(keyFile) > 0 {
keyBytes, err := fileutil.LoadKey(keyFile)
errs = append(errs, err)
key = string(keyBytes)
}
if len(caCertFile) > 0 {
caCertBytes, err := fileutil.LoadCert(caCertFile)
errs = append(errs, err)
caCert = string(caCertBytes)
}
if len(destCACertFile) > 0 {
destCACertBytes, err := fileutil.LoadCert(destCACertFile)
errs = append(errs, err)
destCACert = string(destCACertBytes)
}
```

Now that I am relatively seasoned in Go, the second form skeeves me out.  I
know that the pattern is to check for nil, why isn't this checking for
nil?  Oh it's because you have a too-clever aggregation step down at the
bottom/.  It just doesn't feel right - nil is not an error, don't stick it
in the error list.  Besides that, this eliminates some error-handling
boilerplate.  If we thought that was important we might not be using go.

I don't think we should merge this, sorry.   I'll appeal to @lavalamp for a
third-party opinion, if he cares to offer one.

On Thu, Dec 3, 2015 at 3:06 AM, Michail Kargakis <notifications@github.com>
wrote:

> In pkg/util/errors/errors.go
> <https://github.com/kubernetes/kubernetes/pull/17872#discussion_r46538530>
> :
>
> > @@ -29,10 +29,16 @@ type Aggregate interface {
> >  // is itself an implementation of the error interface.  If the slice is empty,
> >  // this returns nil.
> >  func NewAggregate(errlist []error) Aggregate {
> > -	if len(errlist) == 0 {
> > +	errs := []error{}
> > +	for _, err := range errlist {
> > +		if err != nil {
>
> The last place I felt the need for this functionality was here
> <https://github.com/kargakis/origin/blob/578d92cc58b827d930aeb59027dceb0273a6320e/pkg/route/generator/generate.go#L166>.
> Generally, this if err != nil pattern is disturbing when you don't have
> more than one way to handle the error and I'm happy when I can skip it. It
> may or may not feel Go-like, from one pov even nil is considered a value,
> OTOH what should one do when they have to handle a errors.New("nil") error?
> Special case it? Start parsing errors?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17872/files#r46538530>.
>